code/datums/beam.dm

```diff
@@ -26,53 +27,76 @@
 	icon_state = beam_icon_state
 	beam_type = btype
 
+
 /datum/beam/proc/Start()
 	Draw()
-	while(!finished && target && world.time<endtime && get_dist(origin,target)<max_distance && origin.z == target.z)
+	while(!finished && origin && target && world.time < endtime && get_dist(origin,target)<max_distance && origin.z == target.z)
 		if(origin.loc != origin_oldloc || target.loc != target_oldloc)
+			origin_oldloc = origin.loc //so we don't keep checking against their initial positions, leading to endless Reset()+Draw() calls
+			target_oldloc = target.loc
 			Reset()
 			Draw()
 		sleep(sleep_time)
+
 	qdel(src)
 
+
 /datum/beam/proc/End()
 	finished = 1
 
+
 /datum/beam/proc/Reset()
 	for(var/obj/effect/ebeam/B in elements)
 		qdel(B)
 
+
 /datum/beam/Destroy()
 	Reset()
 	target = null
 	origin = null
 	return ..()
 
+
 /datum/beam/proc/Draw()
-	var/Angle=round(Get_Angle(origin,target))
+	var/Angle = round(Get_Angle(origin,target))
```

Comment:
messing with comment chronology cuz I'm a dum dum

isn't there a helper for this?